server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/InheritanceAction.java

```diff
@@ -75,54 +79,42 @@ public void handle(Request request, Response response) throws Exception {
+    return InheritanceWsResponse.newBuilder()
+      .setProfile(buildProfile(profile, statistics))
+      .addAllAncestors(buildAncestors(ancestors, statistics))
+      .addAllChildren(buildChildren(children, statistics))
+      .build();
   }
 
-  private static void writeProfile(JsonWriter json, QualityProfileDto profile, Statistics statistics) {
-    String profileKey = profile.getKey();
-    json.name("profile");
-    writeProfileAttributes(json, profileKey, profile.getName(), profile.getParentKee(), statistics);
+  private static QualityProfile buildProfile(QualityProfileDto profile, Statistics statistics) {
+    return buildProfile(profile.getKey(), profile.getName(), profile.getParentKee(), statistics);
   }
 
-  private static void writeAncestors(JsonWriter json, List<QualityProfileDto> ancestors, Statistics statistics) {
-    json.name("ancestors").beginArray();
-    for (QualityProfileDto ancestor : ancestors) {
-      String ancestorKey = ancestor.getKey();
-      writeProfileAttributes(json, ancestorKey, ancestor.getName(), ancestor.getParentKee(), statistics);
-    }
-    json.endArray();
-  }
-
-  private static void writeChildren(JsonWriter json, List<QualityProfileDto> children, Statistics statistics) {
-    json.name("children").beginArray();
-    for (QualityProfileDto child : children) {
-      String childKey = child.getKey();
-      writeProfileAttributes(json, childKey, child.getName(), null, statistics);
-    }
-    json.endArray();
+  private static Iterable<QualityProfile> buildAncestors(List<QualityProfileDto> ancestors, Statistics statistics) {
+    return ancestors.stream()
+      .map(ancestor -> buildProfile(ancestor.getKey(), ancestor.getName(), ancestor.getParentKee(), statistics))
+      .collect(Collectors.toList());
   }
 
-  private static void writeProfileAttributes(JsonWriter json, String key, String name, @Nullable String parentKey, Statistics statistics) {
-    json.beginObject();
-    json.prop("key", key)
-      .prop("name", name)
-      .prop("parent", parentKey);
-    writeStats(json, key, statistics);
-    json.endObject();
+  private static Iterable<QualityProfile> buildChildren(List<QualityProfileDto> children, Statistics statistics) {
+    return children.stream()
+      .map(child -> buildProfile(child.getKey(), child.getName(), null, statistics))
+      .collect(Collectors.toList());
   }
 
-  private static void writeStats(JsonWriter json, String profileKey, Statistics statistics) {
-    json.prop("activeRuleCount", statistics.countRulesByProfileKey.getOrDefault(profileKey, 0L));
-    json.prop("overridingRuleCount", statistics.countOverridingRulesByProfileKey.getOrDefault(profileKey, 0L));
+  private static QualityProfile buildProfile(String key, String name, @Nullable String parentKey, Statistics statistics) {
+    return QualityProfile.newBuilder()
+      .setKey(key)
+      .setName(name)
+      .setParent(parentKey == null ? "" : parentKey)
```

Comment:
That changes API specification. The parent field should not be returned when there is not parent. Your change returns `"parent": ""`. A solution is to use:
```
QualityProfile.Builder builder = QualityProfile.newBuilder()
      .setKey(key)
      .setName(name)
      .setActiveRuleCount(statistics.getActiveRuleCount(key))
      .setOverridingRuleCount(statistics.getOverridingRuleCount(key));
Protobuf.setNullable(parentKey, builder::setParent);
return builder.build();
```